<a href="https://colab.research.google.com/github/TRANTHANHDAT94/python_machine_learning/blob/main/Lecture_03_inclass.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Machine Learning Model - Feature Engineering (Tổng hợp biến phái sinh)

In [1]:
# Ví dụ dữ liệu Nominal và Ordinal
# Trình độ học vấn
arr = ['cấp 1', 'cấp 2', 'cấp 3', 'đại học', 'sau đại học'] # => sự sắp xếp nhất định
arr1 = ['nhân viên', 'nhóm trưởng', 'trưởng phòng', 'giám đốc chi nhánh', 'giám đốc vùng'] # => thứ tự nhất định
# Ordinal
# Nominal dữ liệu chữ nhưng ko có giá trị để sắp xếp
arr2 = ['bò sát', 'có vú', 'gia cầm', 'giáp xác']

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [3]:
titanic = sns.load_dataset('titanic')
display(titanic.info(), titanic.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.7+ KB


None

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

## 1.1. Check Data null

In [4]:
titanic.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

**<h2>So you can see that:</h2>** 
- we have age: 177 null 
- we have embark_town 2 null
- ưe have deck 688 null 
- Our task now: how to fulfill data for this null using EDA also

## 1.2. Xu ly du lieu bi thieu(Processing the null value)

In this section we will try to remove the null value by using a good value

In [7]:
titanic['embark_town'].fillna('Unknown')

0      Southampton
1        Cherbourg
2      Southampton
3      Southampton
4      Southampton
          ...     
886    Southampton
887    Southampton
888    Southampton
889      Cherbourg
890     Queenstown
Name: embark_town, Length: 891, dtype: object

In [8]:
titanic['embark_town'] = titanic['embark_town'].fillna('Unknown')#Update titanic['embark_town']
#If we do know well how to fillna well all value will be defined Unknown as defaults

In [9]:
titanic['embark_town'].value_counts()

Southampton    644
Cherbourg      168
Queenstown      77
Unknown          2
Name: embark_town, dtype: int64

In [ ]:
# Fillna for deck

In [ ]:
# Special case: fillna for age:

In [ ]:
titanic['age']

In [ ]:
sns.histplot(titanic, x='age', kde=True)

In [ ]:
titanic['age'].isnull().sum()/len(titanic['age'])

In [ ]:
titanic.groupby(['who', 'sibsp', 'parch'])['age'].agg(lambda x: (x.isnull().sum())/len(x))

In [ ]:
titanic.groupby(['who', 'sibsp', 'parch'])['age'].agg([len, 'mean', lambda x: (x.isnull().sum())/len(x)])

In [ ]:
titanic['age'] = titanic.groupby(['who', 'sibsp', 'parch'])['age'].apply(lambda x: x.fillna(x.mean()))
titanic['age']

In [ ]:
titanic['age'].isnull().sum()

In [ ]:
titanic.groupby(['sibsp', 'parch'])['age'].agg([len, 'mean', lambda x: (x.isnull().sum())/len(x)])

In [ ]:
titanic['age'] = titanic.groupby(['sibsp', 'parch'])['age'].apply(lambda x: x.fillna(x.mean()))
titanic['age']

In [ ]:
titanic['age'].isnull().sum()

In [ ]:
# Dien du lieu bi thieu bang tep sisp va parch lon nhat

In [ ]:
# Tao bien bang cach chia bin du lieu

In [ ]:
# Tao bien bang bien chi bao (female & first class)

In [ ]:
# Tao bien bang tinh z-score (bien doi phan phoi)

In [ ]:
## Data Sampling

In [ ]:
# Tinh ti le survived

In [ ]:
# Random sampling

In [ ]:
# Sampling vs Bootstrap

In [ ]:
# Sampling bang scikit learn